### PR TITLE
add global config parameter for HTTP2 max streams per connection

### PR DESCRIPTION
**Pull request by antoniofastly**

### CODE REVIEW

**Comment by antoniofastly** [comment]
@jhatala @kazuho would appreciate your review.

The boringssl failure looks like a flaky test.  Not sure how to trigger a retry.

**Comment by jhatala** [comment]
> The boringssl failure looks like a flaky test. Not sure how to trigger a retry.

I have triggered a retry on the one failed workflow.

File: lib/http2/connection.c [hunk 1]
```diff
@@ -1961,7 +1954,7 @@ int h2o_http2_handle_upgrade(h2o_req_t *req, struct timeval connected_at)
     req->res.status = 101;
     req->res.reason = "Switching Protocols";
     h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_UPGRADE, NULL, H2O_STRLIT("h2c"));
-    h2o_http1_upgrade(req, (h2o_iovec_t *)&SERVER_PREFACE, 1, on_upgrade_complete, http2conn);
+    h2o_http1_upgrade(req, NULL, 0, on_upgrade_complete, http2conn);
```
Comment (jhatala) [inline lib/http2/connection.c:1957]: Isn't this a regression for h2c?

**Review by jhatala — APPROVED** [review]
Thank you!

File: include/h2o/http2_common.h [hunk 2]
```diff
@@ -175,6 +175,11 @@ typedef struct st_h2o_http2_window_update_payload_t {
     uint32_t window_size_increment;
 } h2o_http2_window_update_payload_t;
 
+typedef struct st_h2o_settings_kvpair {
+    uint16_t key;
+    uint32_t value;
+} h2o_settings_kvpair_t;
```
Comment (kazuho) [inline include/h2o/http2_common.h:181]: Maybe add "http2" to the type name? Use of uint16 for names and uint32 for values are specific to HTTP/2.

File: t/80http2-stream-limit.t [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+# read logs which is an array of [status_code, body_size, stream_id]
+my @logs = do {
+    open my $fh, "<", "$tempdir/access_log"
+        or die "failed to open file:$tempdir/access_log:$!";
+    map { chomp $_; [split / /, $_] } <$fh>;
+};
+debug(join "\n", map { join " ", @$_ } @logs);
+
+cmp_ok scalar(grep { $_->[0] == 200 } @logs), "==", 1000, "accepted all requests";
```
Comment (kazuho) [inline t/80http2-stream-limit.t:97]:
```suggestion
is scalar(grep { $_->[0] == 200 } @logs), 1000, "accepted all requests";
```

File: t/80http2-stream-limit.t [hunk 4]
```diff
@@ -0,0 +1,99 @@
+use strict;
```
Comment (kazuho) [inline t/80http2-stream-limit.t:1]: Please rename the file to t/40http2-stream-limit.t, as 40 is the group we use for testing protocol implementations (80 are tests related to specific bugs).

**Review by kazuho — COMMENTED** [review]
Looks awesome! Code looks great, and the tests are perfect 💯 !! I only have nitpicks.

File: include/h2o/http2_common.h (hunk 2 above)
Comment (antoniofastly) [reply include/h2o/http2_common.h]: Done.

File: t/80http2-stream-limit.t (hunk 3 above)
Comment (antoniofastly) [reply t/80http2-stream-limit.t]: Done.

File: t/80http2-stream-limit.t (hunk 4 above)
Comment (antoniofastly) [reply t/80http2-stream-limit.t]: Done.

**Review by antoniofastly — COMMENTED** [review]
Thanks for the review!

PTAL when you have a chance.

**Comment by kazuho** [comment]
Thank you for the changes! Looks good, will merge once the CI concludes successfully.

**Pull request merged**